binding.cc

```diff
@@ -2992,89 +3152,96 @@ typedef struct sn_crypto_stream_chacha20_ietf_xor_state {
   uint64_t block_counter;
 } sn_crypto_stream_chacha20_ietf_xor_state;
 
-js_value_t *
-sn_crypto_stream_chacha20_ietf_xor_wrap_init (js_env_t *env, js_callback_info_t *info) {
-  SN_ARGV(3, crypto_stream_chacha20_ietf_xor_wrap_init)
-
-  SN_ARGV_BUFFER_CAST(sn_crypto_stream_chacha20_ietf_xor_state *, state, 0)
-  SN_ARGV_TYPEDARRAY(n, 1)
-  SN_ARGV_TYPEDARRAY(k, 2)
-
-  SN_THROWS(state_size != sizeof(sn_crypto_stream_chacha20_ietf_xor_state), "state must be 'crypto_stream_chacha20_ietf_xor_STATEBYTES' bytes")
-  SN_ASSERT_LENGTH(n_size, crypto_stream_chacha20_ietf_NONCEBYTES, "n")
-  SN_ASSERT_LENGTH(k_size, crypto_stream_chacha20_ietf_KEYBYTES, "k")
-
-  state->remainder = 0;
-  state->block_counter = 0;
-  memcpy(state->n, n_data, crypto_stream_chacha20_ietf_NONCEBYTES);
-  memcpy(state->k, k_data, crypto_stream_chacha20_ietf_KEYBYTES);
+static inline void
+sn_crypto_stream_chacha20_ietf_xor_wrap_init(
+  js_env_t *,
+  js_receiver_t,
+  js_typedarray_span_t<> state,
+  js_typedarray_span_t<> n,
+  js_typedarray_span_t<> k
+) {
+  assert(state.size_bytes() == sizeof(sn_crypto_stream_chacha20_ietf_xor_state));
+  assert(n.size_bytes() == crypto_stream_chacha20_ietf_NONCEBYTES);
+  assert(k.size_bytes() == crypto_stream_chacha20_ietf_KEYBYTES);
 
-  return NULL;
+  auto state_data = reinterpret_cast<sn_crypto_stream_chacha20_ietf_xor_state *>(state.data());
+  state_data->remainder = 0;
+  state_data->block_counter = 0;
+  memcpy(state_data->n, n.data(), crypto_stream_chacha20_ietf_NONCEBYTES);
+  memcpy(state_data->k, k.data(), crypto_stream_chacha20_ietf_KEYBYTES);
 }
 
-js_value_t *
-sn_crypto_stream_chacha20_ietf_xor_wrap_update (js_env_t *env, js_callback_info_t *info) {
-  SN_ARGV(3, crypto_stream_chacha20_ietf_xor_wrap_update)
-
-  SN_ARGV_BUFFER_CAST(sn_crypto_stream_chacha20_ietf_xor_state *, state, 0)
-  SN_ARGV_BUFFER_CAST(unsigned char *, c, 1)
-  SN_ARGV_BUFFER_CAST(unsigned char *, m, 2)
+static inline void
+sn_crypto_stream_chacha20_ietf_xor_wrap_update(
+  js_env_t *,
+  js_receiver_t,
+  js_typedarray_span_t<> state,
+  js_typedarray_span_t<> c,
+  js_typedarray_span_t<> m
+) {
+  assert(state.size_bytes() == sizeof(sn_crypto_stream_chacha20_ietf_xor_state));
+  assert(c.size_bytes() == m.size_bytes());
 
-  SN_THROWS(state_size != sizeof(sn_crypto_stream_chacha20_ietf_xor_state), "state must be 'crypto_stream_chacha20_ietf_xor_STATEBYTES' bytes")
-  SN_THROWS(c_size != m_size, "c must be 'm.byteLength' bytes")
+  auto state_data = reinterpret_cast<sn_crypto_stream_chacha20_ietf_xor_state *>(state.data());
```

Comment:
This could be improved by `libjstl` adding `js_typedarray_span_of_t<T>` similar to `js_arraybuffer_span_of_t<T>`.